timemachine/fe/mcgregor.py

```diff
@@ -303,242 +396,161 @@ def mcs(
-        total_leaves_visited += mcs_result.leaves_visited
-
-        if len(mcs_result.all_maps) > 0:
-            # If we timed out but got cores, throw a warning
-            if mcs_result.timed_out and len(mcs_result.all_maps) < max_cores:
-                warnings.warn(
-                    f"Inexhaustive search: reached max number of visits ({max_visits}) and found only "
-                    f"{len(mcs_result.all_maps)} out of {max_cores} desired cores.",
-                    MaxVisitsWarning,
-                )
-            # don't remove this comment and the one below, useful for debugging!
-            # print(
-            # f"==SUCCESS==[NODES VISITED {mcs_result.nodes_visited} | CORE_SIZE {len([x != UNMAPPED for x in mcs_result.all_maps[0]])} | NUM_CORES {len(mcs_result.all_maps)} | NUM_EDGES {mcs_result.num_edges} | time taken: {time.time()-start_time} | time out? {mcs_result.timed_out}]====="
-            # )
-            break
-        elif mcs_result.timed_out:
-            # If timed out, either due to max_visits or max_cores, raise exception.
-            raise NoMappingError(
-                f"Exceeded max number of visits/cores - no valid cores could be found: {total_nodes_visited} nodes visited."
+    mcs_result = MCSResult.from_nodes(nodes, cached_leaf_filter_fxn, max_visits, max_cores)
+
+    if len(mcs_result.all_maps) > 0:
+        # If we timed out but got cores, throw a warning
+        if mcs_result.timed_out and len(mcs_result.all_maps) < max_cores:
+            warnings.warn(
+                f"Inexhaustive search: reached max number of visits ({max_visits}) and found only "
+                + f"{len(mcs_result.all_maps)} out of {max_cores} desired cores.",
+                MaxVisitsWarning,
             )
-        # else:
+        # don't remove this comment and the one below, useful for debugging!
         # print(
-        # f"==FAILED==[NODES VISITED {mcs_result.nodes_visited} | time taken: {time.time()-start_time} | time out? {mcs_result.timed_out}]====="
+        # f"==SUCCESS==[NODES VISITED {mcs_result.nodes_visited} | CORE_SIZE {len([x != UNMAPPED for x in mcs_result.all_maps[0]])} | NUM_CORES {len(mcs_result.all_maps)} | NUM_EDGES {mcs_result.num_edges} | time taken: {time.time()-start_time} | time out? {mcs_result.timed_out}]====="
         # )
 
+    elif mcs_result.timed_out:
+        # If timed out, either due to max_visits or max_cores, raise exception.
+        raise NoMappingError(
+            f"Exceeded max number of visits/cores - no valid cores could be found: {mcs_result.nodes_visited} nodes visited."
+        )
+    # else:
+    # print(
+    # f"==FAILED==[NODES VISITED {mcs_result.nodes_visited} | time taken: {time.time()-start_time} | time out? {mcs_result.timed_out}]====="
+    # )
+
     if len(mcs_result.all_maps) == 0:
-        raise NoMappingError("Unable to find mapping")
+        raise NoMappingError(f"Unable to find mapping with at least {min_num_edges} edges")
```

Comment:
Hm - this looks like it removes the "fast termination" feature we had before when we specified a minimum value for the required # of edges. (i.e. before we can quickly exit in the outer layer, now we only exit *after* the search is complete).

```
if cur_threshold < min_threshold:
    raise NoMappingError(f"Unable to find mapping with at least {min_threshold} edges")
```

IIRC @badisa used this feature for graph generation at some point - is it still being used?

Reply:
If I understand correctly, the early termination behavior is roughly the same (note that `min_threshold` was renamed `min_num_edges`). Because `best_num_edges` is initialized to `min_num_edges` (https://github.com/mcwitt/timemachine/blob/3137fad307d23ba70f83b265f17bcfbd711b85b8/timemachine/fe/mcgregor.py#L454), the search is pruned in the same way it would be during the initial iteration of DFS in the existing code, so I don't think the changes here result in any more nodes being searched in this case.

Reply:
> the search is pruned in the same way it would be during the initial iteration of DFS in the existing code

Hmm, thinking about it more, that's not quite right. It seems like the behavior in this PR is to search _fewer_ nodes than the current version in the case where `min_num_edges` rules out any possible mappings. This is because in this PR we always prune according to `min_num_edges`, versus the current behavior that iterates down from `min(num_edges_a, num_edges_b)` to `min_num_edges`. Does that make sense?

Reply:
ah, yes, I think that makes sense. Thank you for the clarification.